domain/user/state/state.go

```diff
@@ -0,0 +1,395 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/user"
+	"github.com/juju/juju/domain"
+	usererrors "github.com/juju/juju/domain/user/errors"
+	databaseutils "github.com/juju/juju/internal/database"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// AddUser will add a new user to the database. If the user already exists
+// an error that satisfies usererrors.AlreadyExists will be returned. If the
+// users creator is set and does not exist then an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUser(ctx context.Context, uuid user.UUID, user user.User, creatorUuid user.UUID) error {
+	err := uuid.Validate()
```

Comment:
We haven't traditionally been doing double validation as that is the responsibility of the service layer. The state layer doesn't care if it is a valid UUID or not. It cares that the data value we are going to persist is valid for the DDL of the table.

I would only run validation checks here if it is related to the database DDL but SQL is going todo that for us anyway so I don't think they're needed IMO.

Reply:
Same comment for all the checks below.